modules/core/shared/src/main/scala/coursier/core/VersionConstraint.scala

```diff
@@ -61,4 +61,21 @@ object VersionConstraint {
 
     constraintOpt.filter(_.isValid)
   }
+
+  def relaxedMerge(constraints: VersionConstraint*): Option[VersionConstraint] = {
+    merge(constraints: _*) orElse {
+      val nonZeroIntervals = constraints.flatMap(c =>
+        if (c.interval == VersionInterval.zero) Nil
+        else List(c.interval)
+      )
+      val lowerBounds = nonZeroIntervals.flatMap(c =>
+        if (c.fromIncluded) c.from.toList
+        else Nil
+      )
+      val preferreds = constraints.flatMap(_.preferred).toList
+      val vs = (lowerBounds ++ preferreds).distinct
+      if (vs.isEmpty) None
+      else Some(VersionConstraint.preferred(vs.max))
+    }
```

Comment:
I'd like to propose a slightly different algorithm here, more consistent with the default one:
- sort constraints in ascending order in one way or another (using the lower bound of the intervals to compare them with specific versions and other intervals)
- from the right, merge them two-by-two with the `merge` method above, and stop (right before) `merge` returns `None`
- return the last successful merge (or the last element in the sorted list if no `merge` succeeded)

That way, relaxed merge will still behave almost like the other merge, only ignoring things that needs to be ignored. For example, if passed `[1.0,2.0)`, `[3.0,4.0)`, and `2.8`, `[1.0,2.0)` will be ignored, and it'll return `[3.0,4.0)` instead of just `3.0` with the current implementation (down the line, `[3.0,4.0)` will select the highest version in the interval).

Reply:
Addressed in https://github.com/coursier/coursier/pull/1293/commits/bcff24b822d54f3aaba87e627388c8b1a01c5057